py/objtype.c

```diff
@@ -1103,6 +1103,13 @@ mp_obj_t mp_obj_new_type(qstr name, mp_obj_t bases_tuple, mp_obj_t locals_dict)
     if (!mp_obj_is_type(bases_tuple, &mp_type_tuple)) {
         mp_raise_TypeError(NULL);
     }
+    #if MICROPY_PY_COLLECTIONS_ORDEREDDICT
+    if (mp_obj_is_type(locals_dict, &mp_type_ordereddict)) {
+        locals_dict = mp_obj_dict_copy(locals_dict);
+        mp_obj_dict_t *locals = MP_OBJ_TO_PTR(locals_dict);
+        locals->base.type = &mp_type_dict;
+    }
+    #endif
```

Comment:
Instead of making a copy of the dict, I think it'd be much simpler and more efficient to just replace all checks in this file for `locals_dict` being an `mp_type_dict` to test for either `mp_type_dict` or `mp_type_ordereddict`.  It looks like there are only 3 such checks, 2 of which are just assertions.  Probably just add a macro in this file like `IS_LOCALS_COMPATIBLE_DICT(locals_dict)`.

Reply:
I initially thought of this, however at a quick search I thought it looked like other parts of the codebase also expect locals to be a `mp_type_dict`. I'm not sure where I was looking now though, I guess I could try it out and test if it seems to work.

Reply:
I think it'll work... one way to test is to change all dicts that come in here to `mp_type_ordereddict` and then run the test suite.

Reply:
I tested this out, by forcing `locals_dict->base.type = &mp_type_ordereddict` in the code here, and all tests pass on unix coverage build.  So I'd say go for this approach (I can make a PR for this part, just let me know either way).

Reply:
See #6176